### PR TITLE
Cleanup host dependencies ( tentative demo)

### PR DESCRIPTION
**Pull request by otherview**
### Why this change is needed

The host has been decoupled (thank god) and a bunch of services and libs have derived from it.
All of these components have been moved to a different package, which means that circular imports and the existing dependencies are now harder to handle.

This PR cleans up the new Packages into Components or Services. The Services has a `state`, `Start()`, `Stop()`, `Health()`, the Component does not.
This PR also splits between Packages that are of the responsibility of the Host and of the Host Container. `RPC`, `P2P`,`metricsService`, `profiller` are services that are agostic to the host funcionality. `L1Repo`, `L2Repo`, `logEventService`, `enclaveService` are of the Host responsability.

A simple way to determine responsibility is asking the question, can the Host run if the Package is off ?
`RPC`, `P2P`, `metricsService`, `profiller` -> The host will not receive data ( in p2p case there's a todo because the p2p is server and client)
`L1Repo`, `L2Repo`, `logEventService`, `enclaveService`  -> If any of these are off the host crashes.

### CODE REVIEW

File: go/host/rpc/clientapi/client_api_filter.go
Comment (BedrockSquirrel) [inline go/host/rpc/clientapi/client_api_filter.go:1]: I think one potential improvement would be to have these guys define their own interface. At the moment they depend on the whole Host interface which is kinda heavy, would make it very painful to test in isolation.

ATM the host interface is being defined in terms of DB and enclave when it feels like really it should be defined by the components that want to call it.

For example something like this would make this easier to test and perhaps give a clearer contract for the host to work against.

```
type hostFilterService interface {
   Subscribe(id rpc.ID, encryptedLogSubscription common.EncryptedParamsLogSubscription, matchedLogs chan []byte) error
   Unsubscribe(id rpc.ID)
   
   GetLogs(encryptedParams EncryptedParamsGetLogs) (*responses.Logs, SystemError)
}

func NewFilterAPI(host hostFilterService, logger gethlog.Logger) *FilterAPI {
	return &FilterAPI{
		host:   host,
		logger: logger,
	}
}
```

**Review by BedrockSquirrel — COMMENTED** [review]
Few first thoughts but at a high level I'm not against going with this approach. I think it helps achieve the separation we're looking for if you can get it to work.

File: go/host/host.go [hunk 1]
```diff
@@ -13,112 +14,101 @@ import (
...
-		if err != nil {
-			logger.Crit("unable to start the profiler: %s", log.ErrKey, err)
-		}
-	}
 
+	// create the Host controlled Services
+	l1RepoService := l1.NewL1Repository(l1Client, logger)
+	l2RepoService := l2.NewBatchRepository(config, p2p, database, logger)
+	logEventService := events.NewLogEventManager(enclaveClient, logger)
+	// todo review why there are 2 inits for the guardian ( guardian component + service )
```
[the remark this answers is not in the document]
Comment (otherview) [reply go/host/host.go:72]: What if the EnclaveService initialized the Guardian ?

File: go/host/host.go [hunk 2]
```diff
@@ -13,112 +14,101 @@ import (
...
+	enclaveService := enclave.NewService(hostIdentity, p2p, enclaveGuardianLib, logger)
+
+	// log startup data for sanity control / tests
 	jsonConfig, _ := json.MarshalIndent(config, "", "  ")
 	logger.Info("Host service created with following config:", log.CfgKey, string(jsonConfig))
 
-	return host
+	return &host{
+		APIDBRepository:  database,
+		APIEnclaveClient: enclaveClient,
```
[the remark this answers is not in the document]
Comment (otherview) [reply go/host/host.go:95]: > So currently this means the APIs are talking directly to the enclave. 
Yes, that avoids that host impl has to be a passthrough like the enclave impl.

> ut you're proposing I guess that eventually EnclaveService would implement APIEnclaveClient and it can delegate the methods to whichever enclave it wants to?

The proposal would be using this pattern and when we need the "host-hop" we just overload the method. Does that make sense ?

File: go/host/host.go [hunk 3]
```diff
@@ -13,112 +14,101 @@ import (
...
-
-		logger:         logger,
-		metricRegistry: regMetrics,
 
-		stopControl: stopcontrol.New(),
-	}
-	host.interrupter = breaker.Multiplex(
+	// set up control mechanisms
+	stopControl := stopcontrol.New()
+	interrupter := breaker.Multiplex(
```
[the remark this answers is not in the document]
Comment (otherview) [reply go/host/host.go:61]: Wait, the host_container uses the stop, not the interrupt, right ? Otherwise it's doing 2 things.. Or am I getting this wrong ?

File: go/host/enclave/guardian.go [hunk 4]
```diff
@@ -358,7 +380,24 @@ func (g *Guardian) catchupWithL2() error {
 		nextHead := prevHead.Add(prevHead, big.NewInt(1))
 
 		g.logger.Trace("fetching next batch", log.BatchSeqNoKey, nextHead)
-		batch, err := g.sl.L2Repo().FetchBatchBySeqNo(nextHead)
+		batch, err := g.l2Repo.FetchBatchBySeqNo(nextHead)
+		// if data is expected to exist then fetch it from the enclave if it's the sequencer
+		// validators will fetch the data async
+		if g.isSequencer && errors.Is(err, errutil.ErrExpectedData) {
+			// sequencer does not request batches from peers, it checks if its enclave has the batch
+			b, err := g.LookupBatchBySeqNo(nextHead)
```
[the remark this answers is not in the document]
Comment (otherview) [reply go/host/enclave/guardian.go:388]: Yeah good point, I though it made it clearer, but it's probably redudant huh ?

**Comment by otherview** [comment]
Passed 2e2 at https://github.com/obscuronet/obscuro-test/actions/runs/5823333237

**Comment by tudor-malene** [comment]
> This PR cleans up the new Packages into Components or Services. The Services has a `state`, `Start()`, `Stop()`, `Health()`, the Component does not. This PR also splits between Packages that are of the responsibility of the Host and of the Host Container. `RPC`, `P2P`,`metricsService`, `profiller` are services that are agostic to the host funcionality. `L1Repo`, `L2Repo`, `logEventService`, `enclaveService` are of the Host responsability.
> 
> A simple way to determine responsibility is asking the question, can the Host run if the Package is off ? `RPC`, `P2P`, `metricsService`, `profiller` -> The host will not receive data ( in p2p case there's a todo because the p2p is server and client) `L1Repo`, `L2Repo`, `logEventService`, `enclaveService` -> If any of these are off the host crashes.

I'm not convinced about this rule. It's a good rule, but not sure if it's the most appropriate one to apply in this situation

File: go/common/host/services.go [hunk 5]
```diff
@@ -111,8 +95,6 @@ type L1Publisher interface {
 	// PublishSecretResponse will create and publish a secret response tx to the management contract - fire and forget we don't wait for receipt
 	PublishSecretResponse(secretResponse *common.ProducedSecretResponse) error
```
Comment (tudor-malene) [inline go/common/host/services.go:96]: I don't quite get it.
Isn't "Service" supposed to be a generic interface implemented by all Services that a host depends on?
Or is this approach different?

**Review by tudor-malene — COMMENTED** [review]
I find it quite hard to follow what's going on here. 

It would help to have a list with each component (or service), what responsibilities it has, and how they're wired together
